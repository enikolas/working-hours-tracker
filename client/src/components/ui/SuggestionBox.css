ul.suggestion-box {
  position: absolute;
  background: #fff;
  list-style: none;
  border: 2px solid #f93;
  box-shadow: 2px 2px 8px rgba(0,0,0,0.4);
  margin: 45px 0 0 70px;
  padding: 0;
  width: auto;
  display: flex;
  flex-direction: row;
  font-size: 12px;
}

ul.suggestion-box.minutes {
  margin-left: 199px;
}

ul.suggestion-box li {
  display: inline-block;
  margin: 0;
}

ul.suggestion-box li button {
  padding: 5px 5px;
  cursor: pointer;
  display: inline-block;
  font-size: 14px;
  border: 0;
  background: transparent;
}

ul.suggestion-box li button:hover {
  background-color: #777;
  color: #fff;
}
