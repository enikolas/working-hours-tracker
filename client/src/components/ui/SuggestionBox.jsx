import React from 'react';
import PropTypes from 'prop-types';

import './SuggestionBox.css';

const withLeadingZero = number => (number < 10 ? `0${number}` : String(number));

class SuggestionBox extends React.Component {
  constructor(props) {
    super(props);

    if (props.mode === 'minutes') {
      this.valueList = [
        0,
        15,
        30,
        45
      ];
    } else {
      this.valueList = [];
      const refHour = props.referenceHour;
      for (let r = refHour - 2; r <= refHour + 2; r += 1) {
        this.valueList.push(String(r));
      }
    }

    this._actionCall = this._actionCall.bind(this);
    this._individualSuggestion = this._individualSuggestion.bind(this);
  }

  _actionCall(value) {
    return () => {
      const valueToSend = this.mode === 'hours' ? value : withLeadingZero(value);
      this.props.onChoose(valueToSend);
    };
  }

  _individualSuggestion(value) {
    return (
      <li key={value}>
        <button type="button" onClick={this._actionCall(value)} tabIndex="-1">
          {this.mode === 'hours' ? value : withLeadingZero(value)}
        </button>
      </li>
    );
  }

  render() {
    const { show, mode } = this.props;
    if (!show) {
      return null;
    }
    return (
      <ul className={`suggestion-box ${mode}`}>
        {this.valueList.map(this._individualSuggestion)}
      </ul>
    );
  }
}

export default SuggestionBox;

SuggestionBox.propTypes = {
  show: PropTypes.bool,
  mode: PropTypes.oneOf(['minutes', 'hours']).isRequired,
  onChoose: PropTypes.func,
  referenceHour: PropTypes.number
};

SuggestionBox.defaultProps = {
  show: false,
  onChoose: () => {},
  referenceHour: 9
};
