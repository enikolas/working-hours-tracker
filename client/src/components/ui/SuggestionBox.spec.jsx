import React from 'react';
import { shallow } from 'enzyme';

import SuggestionBox from './SuggestionBox';

describe('SuggestionBox', () => {
	describe('render', () => {
		it('should render properly with min props', () => {
			const invisibleWrapper = shallow(<SuggestionBox mode="hours" />);
			const hoursWrapper = shallow(<SuggestionBox show mode="hours" />);
			const minutesWrapper = shallow(<SuggestionBox show mode="minutes" />);
			expect(invisibleWrapper).toMatchSnapshot();
			expect(hoursWrapper).toMatchSnapshot();
			expect(minutesWrapper).toMatchSnapshot();
		});
	});
	describe('onClick', () => {
		it('should call a function with the value selected', () => {
			const onChoose = jest.fn();
			const choosableWrapper = shallow(<SuggestionBox
				show
				mode="minutes"
				onChoose={onChoose}
			/>);

			const fifteenMinutesLi = choosableWrapper.find('li').at(1);
			const minuteSelect = fifteenMinutesLi.find('button');
			minuteSelect.simulate('click', { preventDefault: jest.fn() });

			expect(onChoose).toHaveBeenCalledWith('15');
		});
	});
});
