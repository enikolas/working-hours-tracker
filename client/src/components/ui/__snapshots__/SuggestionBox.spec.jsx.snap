// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestionBox render should render properly with min props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SuggestionBox
    mode="hours"
    onChoose={[Function]}
    referenceHour={9}
    show={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`SuggestionBox render should render properly with min props 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SuggestionBox
    mode="hours"
    onChoose={[Function]}
    referenceHour={9}
    show={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <li>
          <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            07
          </button>
        </li>,
        <li>
          <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            08
          </button>
        </li>,
        <li>
          <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            09
          </button>
        </li>,
        <li>
          <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            10
          </button>
        </li>,
        <li>
          <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            11
          </button>
        </li>,
      ],
      "className": "suggestion-box hours",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "7",
        "nodeType": "host",
        "props": Object {
          "children": <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            07
          </button>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "07",
            "onClick": [Function],
            "tabIndex": "-1",
            "type": "button",
          },
          "ref": null,
          "rendered": "07",
          "type": "button",
        },
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "8",
        "nodeType": "host",
        "props": Object {
          "children": <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            08
          </button>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "08",
            "onClick": [Function],
            "tabIndex": "-1",
            "type": "button",
          },
          "ref": null,
          "rendered": "08",
          "type": "button",
        },
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "9",
        "nodeType": "host",
        "props": Object {
          "children": <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            09
          </button>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "09",
            "onClick": [Function],
            "tabIndex": "-1",
            "type": "button",
          },
          "ref": null,
          "rendered": "09",
          "type": "button",
        },
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "10",
        "nodeType": "host",
        "props": Object {
          "children": <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            10
          </button>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "10",
            "onClick": [Function],
            "tabIndex": "-1",
            "type": "button",
          },
          "ref": null,
          "rendered": "10",
          "type": "button",
        },
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "11",
        "nodeType": "host",
        "props": Object {
          "children": <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            11
          </button>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "11",
            "onClick": [Function],
            "tabIndex": "-1",
            "type": "button",
          },
          "ref": null,
          "rendered": "11",
          "type": "button",
        },
        "type": "li",
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <li>
            <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              07
            </button>
          </li>,
          <li>
            <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              08
            </button>
          </li>,
          <li>
            <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              09
            </button>
          </li>,
          <li>
            <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              10
            </button>
          </li>,
          <li>
            <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              11
            </button>
          </li>,
        ],
        "className": "suggestion-box hours",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "7",
          "nodeType": "host",
          "props": Object {
            "children": <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              07
            </button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "07",
              "onClick": [Function],
              "tabIndex": "-1",
              "type": "button",
            },
            "ref": null,
            "rendered": "07",
            "type": "button",
          },
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "8",
          "nodeType": "host",
          "props": Object {
            "children": <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              08
            </button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "08",
              "onClick": [Function],
              "tabIndex": "-1",
              "type": "button",
            },
            "ref": null,
            "rendered": "08",
            "type": "button",
          },
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "9",
          "nodeType": "host",
          "props": Object {
            "children": <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              09
            </button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "09",
              "onClick": [Function],
              "tabIndex": "-1",
              "type": "button",
            },
            "ref": null,
            "rendered": "09",
            "type": "button",
          },
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "10",
          "nodeType": "host",
          "props": Object {
            "children": <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              10
            </button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "10",
              "onClick": [Function],
              "tabIndex": "-1",
              "type": "button",
            },
            "ref": null,
            "rendered": "10",
            "type": "button",
          },
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "11",
          "nodeType": "host",
          "props": Object {
            "children": <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              11
            </button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "11",
              "onClick": [Function],
              "tabIndex": "-1",
              "type": "button",
            },
            "ref": null,
            "rendered": "11",
            "type": "button",
          },
          "type": "li",
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`SuggestionBox render should render properly with min props 3`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SuggestionBox
    mode="minutes"
    onChoose={[Function]}
    referenceHour={9}
    show={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <li>
          <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            00
          </button>
        </li>,
        <li>
          <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            15
          </button>
        </li>,
        <li>
          <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            30
          </button>
        </li>,
        <li>
          <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            45
          </button>
        </li>,
      ],
      "className": "suggestion-box minutes",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "host",
        "props": Object {
          "children": <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            00
          </button>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "00",
            "onClick": [Function],
            "tabIndex": "-1",
            "type": "button",
          },
          "ref": null,
          "rendered": "00",
          "type": "button",
        },
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "15",
        "nodeType": "host",
        "props": Object {
          "children": <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            15
          </button>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "15",
            "onClick": [Function],
            "tabIndex": "-1",
            "type": "button",
          },
          "ref": null,
          "rendered": "15",
          "type": "button",
        },
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "30",
        "nodeType": "host",
        "props": Object {
          "children": <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            30
          </button>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "30",
            "onClick": [Function],
            "tabIndex": "-1",
            "type": "button",
          },
          "ref": null,
          "rendered": "30",
          "type": "button",
        },
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "45",
        "nodeType": "host",
        "props": Object {
          "children": <button
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          >
            45
          </button>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "45",
            "onClick": [Function],
            "tabIndex": "-1",
            "type": "button",
          },
          "ref": null,
          "rendered": "45",
          "type": "button",
        },
        "type": "li",
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <li>
            <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              00
            </button>
          </li>,
          <li>
            <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              15
            </button>
          </li>,
          <li>
            <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              30
            </button>
          </li>,
          <li>
            <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              45
            </button>
          </li>,
        ],
        "className": "suggestion-box minutes",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "host",
          "props": Object {
            "children": <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              00
            </button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "00",
              "onClick": [Function],
              "tabIndex": "-1",
              "type": "button",
            },
            "ref": null,
            "rendered": "00",
            "type": "button",
          },
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "15",
          "nodeType": "host",
          "props": Object {
            "children": <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              15
            </button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "15",
              "onClick": [Function],
              "tabIndex": "-1",
              "type": "button",
            },
            "ref": null,
            "rendered": "15",
            "type": "button",
          },
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "30",
          "nodeType": "host",
          "props": Object {
            "children": <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              30
            </button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "30",
              "onClick": [Function],
              "tabIndex": "-1",
              "type": "button",
            },
            "ref": null,
            "rendered": "30",
            "type": "button",
          },
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "45",
          "nodeType": "host",
          "props": Object {
            "children": <button
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              45
            </button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "45",
              "onClick": [Function],
              "tabIndex": "-1",
              "type": "button",
            },
            "ref": null,
            "rendered": "45",
            "type": "button",
          },
          "type": "li",
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
